### PR TITLE
Remove support for non-compliant subscription requests and add default implementation to handle ping/pong requests

### PR DESCRIPTION
**Pull request by MohamedSabthar**
## Purpose

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/3608
Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/3943

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

File: ballerina/ping_message_job.bal [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                    check task:unscheduleJob(id);
+                    self.id = ();
+                    return;
+                }
+                PingMessage message = {'type: WS_PING};
+                check writeMessage(self.caller, message);
+            }
+        } on fail error cause {
+            string message = cause is websocket:Error ? "Failed to send ping message"
+                : "Filed to unschedule PingMessageJob";
```
Comment (DimuthuMadushan) [inline ballerina/ping_message_job.bal:58]:
```suggestion
                : "Failed to unschedule PingMessageJob";
```

File: ballerina/pong_message_handler_job.bal [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                    return;
+                }
+                if self.caller.isOpen() && !self.pongReceived {
+                    SubscriptionError err = error("Request timeout", code = 4408);
+                    closeConnection(self.caller, err, timeout = 0);
+                }
+                self.pongReceived = false;
+            }
+        } on fail error cause {
+            string message = "Filed to unschedule PingMessageJob";
```
Comment (DimuthuMadushan) [inline ballerina/pong_message_handler_job.bal:69]:
```suggestion
            string message = "Failed to unschedule PingMessageJob";
```

Comment (MohamedSabthar) [reply ballerina/pong_message_handler_job.bal]: Fixed via [468d3f5](https://github.com/ballerina-platform/module-ballerina-graphql/pull/1170/commits/468d3f5f7af98dbc65a55ac256dde488e3c1644d)

File: ballerina/ping_message_job.bal (hunk 1 above)
Comment (MohamedSabthar) [reply ballerina/ping_message_job.bal]: Fixed via [468d3f5](https://github.com/ballerina-platform/module-ballerina-graphql/pull/1170/commits/468d3f5f7af98dbc65a55ac256dde488e3c1644d)

File: ballerina/websocket_utils.bal [hunk 3]
```diff
@@ -118,43 +99,45 @@ isolated function getSubscriptionResponse(Engine engine, __Schema schema, Contex
...
-        check caller->writeMessage(jsonResponse);
-    } else {
-        check caller->writeMessage(payload);
+isolated function closeConnection(websocket:Caller caller, SubscriptionError cause, decimal timeout = 5) {
+    string reason = cause.message();
+    int statusCode = cause.detail().code;
+    error? closedConnection = caller->close(statusCode, reason, timeout);
+    if closedConnection is error {
+        error err = error("Failed to close WebSocket connection", closedConnection);
+        log:printError(err.message(), stackTrace = err.stackTrace());
```
Comment (DimuthuMadushan) [inline ballerina/websocket_utils.bal:108]: Can't we use `logError` function here?

Comment (MohamedSabthar) [reply ballerina/websocket_utils.bal]: Yes, fixed with [3f4ef7e](https://github.com/ballerina-platform/module-ballerina-graphql/pull/1170/commits/3f4ef7ebef3f3ff29e0b7e484de1bbd6b069a876)

File: ballerina/listener_utils.bal [hunk 4]
```diff
@@ -354,17 +354,17 @@ isolated function getHttpService(Engine gqlEngine, GraphqlServiceConfig? service
...
-            }
+    UpgradeService websocketUpgradeService = @websocket:ServiceConfig {
+        subProtocols: [GRAPHQL_TRANSPORT_WS]
+    } isolated service object {
+        isolated resource function get .(http:Request request) 
+        returns websocket:Service|websocket:UpgradeError|http:HeaderNotFoundError {
+            _ = check request.getHeader(WS_SUB_PROTOCOL);
             Context context = check initContext(gqlEngine, contextInitFunction, request);
-            return new WsService(gqlEngine, schema, customHeaders.cloneReadOnly(), context);
+            return new WsService(gqlEngine, schema, context);
```
Comment (DimuthuMadushan) [inline ballerina/listener_utils.bal:364]: Is there any reason to not support configuring `connectionInitWaitTimeout` that is specified in the [spec](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#message-types)?

Comment (MohamedSabthar) [reply ballerina/listener_utils.bal:364]: Currently the default value from the websocket module is for `connectionInitWaitTimeout`. We can address this in a separate issue

Comment (MohamedSabthar) [reply ballerina/listener_utils.bal:364]: Issue: [#3968](https://github.com/ballerina-platform/ballerina-standard-library/issues/3968)

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**